Avoid duplicate calls in Channel Edit init w/ throttle and memoize

nucleogenesis opened the pull request:
## Description

There are several components where `loadChildren` is called. Many of those components call it in their `mounted` or `created` hooks conditionally - basically asking "Is the data I want in the store? No, then `loadChildren()`!" 

Since the components are checking for data before any other component has successfully retrieved the data - they all will call dispatch the `loadChildren` action within ms of each other.

---

I think we should reevaluate our use of the `created` and/or `mounted` hooks for Vuex store initialization across components. Ideally we would initialize things in one place and make everything else wait to render until it's done.

---

This solution that I worked out instead kind of gives us license to be reckless with our data initialization by using `lodash.throttle && lodash.memoize`.

Roughly speaking - we memoize (cache) the results of a throttled function for as long as it is throttled. This is done because `memoize` uses the first argument given as it's cache key and since it's throttling a vuex action, it gets `context` as the first arg. 

So - anyway - we cache it for `wait` ms and then clear the cache. This results in multiple dispatches to `loadChildren` and `loadContentNodes` _with the same paramters_ to only end up being called once every `wait` ms. After `wait` ms, subsequent calls with the same parameters will be called because the cache is cleared.

Frankly, this needs some tests and might be a sideways or unnecessarily complicated approach with only one real application. Whereas revising how/where we make our initialization call(s) may be a bit more of a refactor. I'll revisit this after getting some feedback and revise accordingly.

#### Issue Addressed (if applicable)

Fixes #1882 

## Steps to Test

- [ ] Before checking this out, go to a channel edit for a channel and watch the network tab in devtools. Watch for duplicate values.
- [ ] Check this PR out and start fresh (clear local cache to be sure)
- [ ] Go to that same channel and see that all of the calls were only made once.
- [ ] Navigate throughout the channel both using the tree view and the main view.

- Any performance issues between the two branches?
- Any errors or issues found while navigating between sections?

## Checklist


*Delete any items that don't apply*

- [ ] Is the code clean and well-commented?
- [ ] Are there tests for this change?

## Reviewers

If you are looking to assign a reviewer, here are some options:
- Jordan jayoshih (full stack)
- Aron aronasorman (back end, devops)
- Micah micahscopes (full stack)
- Kevin kollivier (back end)
- Ivan ivanistheone ([Ricecooker](https://github.com/learningequality/ricecooker))
- Richard rtibbles (full stack, [Kolibri](https://github.com/learningequality/kolibri))
- Radina @radinamatic (documentation)

MisRob commented:
> There are several components where loadChildren is called. Many of those components call it in their mounted or created hooks conditionally - basically asking "Is the data I want in the store? No, then loadChildren()!"

How often do we do that? I am aware of only one problem so far. If it's the case, then I am a bit afraid that this solution will lead to adding more dependencies over time instead of being careful of what to connect the store.

> Frankly, this needs some tests and might be a sideways or unnecessarily complicated approach with only one real application. Whereas revising how/where we make our initialization call(s) may be a bit more of a refactor. 

If only one place is making trouble so far, then I would say that another solution would be to leave the two calls there and lower priority of this task for now. I don't see it as a blocker, or are there some significant performance issues at this point related to the issue? If those two calls are slowing it down, then I would say let's merge and open a follow-up tech debt issue.

MisRob commented:
After some more thinking, I was wondering if we could use this solution when we have it available because in any case it'll save us some calls which is not a bad thing (especially taking into account that in general we have performance issues) but also add some console warning whenever throttle is actually used (more same calls than one in a very short period of time). The warning could say that we don't encourage active use of that functionality, ask a developer to consider checking architecture and could also contain a link to a follow-up issue with more detailed explanation?

That would hopefully prevent us from misusing it when not necessary and allow us to use this solution (nice by the way, could be handy whenever there is no other way - I wasn't aware that something like this is possible to implement in a couple of lines) while also addressing my architectural concerns.

Would that make sense?

rtibbles commented on contentcuration/contentcuration/frontend/channelEdit/vuex/contentNode/actions.js:
```diff
@@ -5,14 +5,50 @@ import client from 'shared/client';
+// as its cache key. The first arg is always vuex context, so every call
+// may just use the same memoized results despite receiving different
+// parameters. We splice(1) to remove context altogether.
+const memoCacheResolver = (...args) => JSON.stringify(args.splice(1));
```
Note that this cache resolver may result in cache misses, due to differing key orders in the parameters object.

To fix this, we either need to be specific about which parameters we want to accept and generate a key based on that, or we need to order the keys of the params object before stringifying.

See here for an example:
https://github.com/learningequality/kolibri/blob/release-v0.14.x/kolibri/core/assets/src/api-resource.js#L607

nucleogenesis replied:
Ooooh great catch. I implemented a fn to do this recursively for nested objects as well.

rtibbles commented on contentcuration/contentcuration/frontend/channelEdit/vuex/contentNode/actions.js:
```diff
@@ -5,14 +5,50 @@ import client from 'shared/client';
+// Adapted with gratitude from @Galadirith's comment: 
+// https://github.com/lodash/lodash/issues/2403#issuecomment-290760787
+function memoizedThrottle(func, wait=0, opts={}) {
+  let memo = memoize(
```
Can be `const` instead of `let`.

nucleogenesis commented:
Thanks @rtibbles & @jayoshih for the review.

I've updated it to recursively sort object keys and reduce the cache key size quickly (stringified objects were like 50kb+). 

I also moved everything to the utils for that vuex module.

If we want this tool to be more widely available for use, then I'll move it to `shared/utils`.

As to @MisRob 's points - I feel the same way. I think this solution is more interesting than it is ideal.  I went down the wrong rabbit hole trying to find a solution for this problem and really just need to disentangle the places where this data is initialized and the conditions that cause it to be called.

I'd suggest keeping this as a draft for now and we can return to it if/when we feel it's necessary to use. I'll hope to have a PR for the linked issue and others today.

rtibbles commented:
I avoided suggesting doing this, as it's a step even further towards generalization, but my instinct is that doing this in the API request layer (as opposed to the Vuex action layer) would give us more of what we want here. Basically, any time a request is made through the resource layer, we can cache the promise for the request for X amount of time, and just immediately return that when a request with the same params is put in.

This has the slight advantage that it would also dedupe requests across multiple browser tabs.

rtibbles commented:
The above strategy may also be better implemented inside a service worker, which is designed for just this kind of request interception, as opposed to the slightly hacky construction that I have in place now.

nucleogenesis commented:
@rtibbles the branch will live on if we ever want to reference it - but I will close this PR for now and address the issues I intended to fix in a separate PR

MisRob commented:
@nucleogenesis Thanks Jacob for all your effort dealing with this, it's quite a tricky issue. I learned new things here :)
